hibernate-cache/wildfly-local/src/main/java/org/infinispan/tutorial/simple/hibernate/cache/wildfly/local/InfinispanHibernateCacheWildflyLocal.java

```diff
@@ -0,0 +1,247 @@
+      printfAssert("Query cache put: %d (expected %d)%n", stats.getQueryCachePutCount(), 1, out);
+
+      // Repeat query, expect:
+      // * two cache hits for the number of entities in cache
+      // * a query cache hit
+      queryEntities(out);
+      stats = getStatistics();
+      printfAssert("Event entity cache hits: %d (expected %d)%n", stats.getSecondLevelCacheHitCount(), 2, out);
+      printfAssert("Query cache hit: %d (expected %d)%n", stats.getQueryCacheHitCount(), 1, out);
+
+      // Update one of the persisted entities, stats should show a cache hit and a cache put
+      updateEntity(2L, out);
+      eventCacheStats = getCacheStatistics(EVENT_REGION_NAME);
+      printfAssert("Event entity cache hits: %d (expected %d)%n", eventCacheStats.getHitCount(), 1, out);
+      printfAssert("Event entity cache puts: %d (expected %d)%n", eventCacheStats.getPutCount(), 1, out);
+
+      // Repeat query after update, expect:
+      // * no cache hits or puts since entities are already cached
+      // * a query cache miss and query cache put, because when an entity is updated,
+      //   any queries for that type are invalidated
+      queryEntities(out);
+      stats = getStatistics();
+      printfAssert("Query cache miss: %d (expected %d)%n", stats.getQueryCacheMissCount(),1, out);
+      printfAssert("Query cache put: %d (expected %d)%n", stats.getQueryCachePutCount(), 1, out);
+
+      return out.toString();
+   }
+
+   @GET
+   @Path("/hibernate-cache/expiring")
+   @Produces("application/json")
+   public String expiringEntityCache() throws Exception {
+      StringBuilder out = new StringBuilder();
+
+      SecondLevelCacheStatistics personCacheStats;
+      
+      // Save cache-expiring entity, stats should show a second level cache put
+      saveExpiringEntity();
+      personCacheStats = getCacheStatistics(PERSON_REGION_NAME);
+      printfAssert("Person entity cache puts: %d (expected %d)%n", personCacheStats.getPutCount(), 1, out);
+
+      // Find expiring entity, stats should show a second level cache hit
+      findExpiringEntity(4L, out);
+      personCacheStats = getCacheStatistics(PERSON_REGION_NAME);
+      printfAssert("Person entity cache hits: %d (expected %d)%n", personCacheStats.getHitCount(), 1, out);
+
+      // Wait long enough for entity to be expired from cache
+      Thread.sleep(1100);
+
+      // Find expiring entity, after expiration entity should come from DB
+      // Stats should show a cache miss and a cache put
+      findExpiringEntity(4L, out);
+      personCacheStats = getCacheStatistics(PERSON_REGION_NAME);
+      printfAssert("Person entity cache miss: %d (expected %d)%n", personCacheStats.getMissCount(), 1, out);
+      printfAssert("Person entity cache put: %d (expected %d)%n", personCacheStats.getPutCount(), 1, out);
+
+      return out.toString();
+   }
+
+   private void persistEntities() throws Exception {
+      clearStatistics();
+      userTransaction.begin();
```

Comment:
After the userTransaction.begin(), you should call em.joinTransaction(), as I don't think that the container managed entity manager will otherwise be associated with the JTA transaction.